File: mobile/ios/Runner/VitalsProcessor.swift
import Foundation
import MediaPipeTasksVision
import AVFoundation
import UIKit

class VitalsProcessor: NSObject, AVCaptureVideoDataOutputSampleBufferDelegate {
    private var faceLandmarker: FaceLandmarker?
    private var isCollecting = false
    private var results: [String: [String: [Double]]] = [:]
    private var geometry: [String: [Double]] = [:]
    private var capturedImageBase64: String?
    private var frameCount = 0

    // リアルタイム通知用コールバック
    var onFrameUpdate: (([String: Any]) -> Void)?

    override init() {
        super.init()
        setupLandmarker()
        resetResults()
    }

    private func setupLandmarker() {
        guard let modelPath = Bundle.main.path(forResource: "face_landmarker", ofType: "task", inDirectory: "Frameworks/App.framework/flutter_assets/assets/models") else {
            print("Failed to find face_landmarker.task in assets")
            return
        }

        let options = FaceLandmarkerOptions()
        options.baseOptions.modelAssetPath = modelPath
        options.runningMode = .video
        options.numFaces = 1

        do {
            faceLandmarker = try FaceLandmarker(options: options)
        } catch {
            print("Failed to initialize FaceLandmarker: \(error)")
        }
    }

    private func resetResults() {
        results = [
            "forehead": ["r": [], "g": [], "b": []],
            "left_cheek": ["r": [], "g": [], "b": []],
            "right_cheek": ["r": [], "g": [], "b": []],
            "under_eye_left": ["r": [], "g": [], "b": []],
            "under_eye_right": ["r": [], "g": [], "b": []],
            "lips": ["r": [], "g": [], "b": []],
            "left_cheek_hollow": ["r": [], "g": [], "b": []],
            "right_cheek_hollow": ["r": [], "g": [], "b": []]
        ]
        geometry = [
            "face_width": [],
            "face_height": [],
            "eye_aperture": []
        ]
    }

    func startCollection() {
        resetResults()
        capturedImageBase64 = nil
        frameCount = 0
        isCollecting = true
    }

    func stopCollection() -> [String: Any] {
        isCollecting = false
        var finalResult: [String: Any] = results
        finalResult["geometry"] = geometry
        if let base64 = capturedImageBase64 {
            finalResult["thumbnail"] = base64
        }
        return finalResult
    }

    func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
        guard isCollecting, let faceLandmarker = faceLandmarker else { return }
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
        
        if frameCount == 0 {
            captureThumbnail(from: pixelBuffer)
        }
        frameCount += 1

        let timestamp = Int(CMSampleBufferGetPresentationTimeStamp(sampleBuffer).seconds * 1000)
        
        do {
            let mpImage = try MPImage(pixelBuffer: pixelBuffer)
            let result = try faceLandmarker.detect(videoFrame: mpImage, timestampInMilliseconds: timestamp)
            
            if let landmarks = result.faceLandmarks.first {
                processLandmarks(landmarks, in: pixelBuffer)
            } else {
                onFrameUpdate?(["status": "not_detected"])
            }
        } catch {
            print("MediaPipe detection error: \(error)")
        }
    }

    private func captureThumbnail(from pixelBuffer: CVPixelBuffer) {
        let ciImage = CIImage(cvPixelBuffer: pixelBuffer)
        let context = CIContext(options: nil)
        if let cgImage = context.createCGImage(ciImage, from: ciImage.extent) {
            let uiImage = UIImage(cgImage: cgImage)
            if let resizedImage = self.resizeImage(image: uiImage, targetWidth: 480) {
                if let data = resizedImage.jpegData(compressionQuality: 0.5) {
                    capturedImageBase64 = data.base64EncodedString()
                }
            }
        }
    }

    private func resizeImage(image: UIImage, targetWidth: CGFloat) -> UIImage? {
        let size = image.size
        let ratio = targetWidth / size.width
        let newSize = CGSize(width: targetWidth, height: size.height * ratio)
        UIGraphicsBeginImageContextWithOptions(newSize, false, 1.0)
        image.draw(in: CGRect(origin: .zero, size: newSize))
        let newImage = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()
        return newImage
    }

    private func processLandmarks(_ landmarks: [NormalizedLandmark], in pixelBuffer: CVPixelBuffer) {
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }
        
        // ROI抽出
        extractRGB(landmark: landmarks[10], name: "forehead", in: pixelBuffer)
        extractRGB(landmark: landmarks[234], name: "left_cheek", in: pixelBuffer)
        extractRGB(landmark: landmarks[454], name: "right_cheek", in: pixelBuffer)
        extractRGB(landmark: landmarks[101], name: "under_eye_left", in: pixelBuffer)
        extractRGB(landmark: landmarks[330], name: "under_eye_right", in: pixelBuffer)
        extractRGB(landmark: landmarks[0], name: "lips", in: pixelBuffer)
        extractRGB(landmark: landmarks[214], name: "left_cheek_hollow", in: pixelBuffer)
        extractRGB(landmark: landmarks[434], name: "right_cheek_hollow", in: pixelBuffer)
        
        // 幾何学的特徴の算出
        let f_width = abs(landmarks[454].x - landmarks[234].x)
        let f_height = abs(landmarks[152].y - landmarks[10].y)
        let leftEyeApt = abs(landmarks[159].y - landmarks[145].y) / max(0.01, abs(landmarks[133].x - landmarks[33].x))
        let rightEyeApt = abs(landmarks[386].y - landmarks[374].y) / max(0.01, abs(landmarks[362].x - landmarks[263].x))
        
        geometry["face_width"]?.append(Double(f_width))
        geometry["face_height"]?.append(Double(f_height))
        geometry["eye_aperture"]?.append(Double((leftEyeApt + rightEyeApt) / 2.0))
        
        // UI通知 (主要な点のみ)
        onFrameUpdate?([
            "status": "detected",
            "landmarks": [
                "forehead": ["x": landmarks[10].x, "y": landmarks[10].y],
                "left_cheek": ["x": landmarks[234].x, "y": landmarks[234].y],
                "right_cheek": ["x": landmarks[454].x, "y": landmarks[454].y],
                "lips": ["x": landmarks[0].x, "y": landmarks[0].y]
            ],
            "geometry": [
                "face_width": f_width,
                "face_height": f_height,
                "eye_aperture": (leftEyeApt + rightEyeApt) / 2.0
            ]
        ])
    }

    private func extractRGB(landmark: NormalizedLandmark, name: String, in pixelBuffer: CVPixelBuffer) {
        let w = CVPixelBufferGetWidth(pixelBuffer)
        let h = CVPixelBufferGetHeight(pixelBuffer)
        let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)
        guard let buffer = CVPixelBufferGetBaseAddress(pixelBuffer)?.assumingMemoryBound(to: UInt8.self) else { return }

        let centerX = Int(landmark.x * Float(w))
        let centerY = Int(landmark.y * Float(h))
        let roiSize = 15
        let halfSize = roiSize / 2
        
        var r: Double = 0, g: Double = 0, b: Double = 0, count: Double = 0
        
        for dy in -halfSize..<halfSize {
            for dx in -halfSize..<halfSize {
                let x = centerX + dx, y = centerY + dy
                if x >= 0 && x < w && y >= 0 && y < h {
                    let offset = y * bytesPerRow + x * 4
                    b += Double(buffer[offset])
                    g += Double(buffer[offset + 1])
                    r += Double(buffer[offset + 2])
                    count += 1
                }
            }
        }
        
        if count > 0 {
            results[name]?["r"]?.append(r / count)
            results[name]?["g"]?.append(g / count)
            results[name]?["b"]?.append(b / count)
        }
    }
}

File: mobile/android/app/src/main/kotlin/com/example/hono_health_mobile/VitalsProcessor.kt
package com.example.hono_health_mobile

import android.content.Context
import android.graphics.Bitmap
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker.FaceLandmarkerOptions
import java.io.ByteArrayOutputStream
import java.util.Base64

class VitalsProcessor(private val context: Context) {
    private var faceLandmarker: FaceLandmarker? = null
    private var isCollecting = false
    private var results = mutableMapOf<String, MutableMap<String, MutableList<Double>>>()
    private var geometry = mutableMapOf<String, MutableList<Double>>()
    private var capturedImageBase64: String? = null
    private var frameCount = 0

    var onFrameUpdate: ((Map<String, Any>) -> Unit)? = null

    init {
        setupLandmarker()
        resetResults()
    }

    private fun setupLandmarker() {
        val baseOptionsBuilder = BaseOptions.builder()
            .setModelAssetPath("face_landmarker.task")

        val optionsBuilder = FaceLandmarkerOptions.builder()
            .setBaseOptions(baseOptionsBuilder.build())
            .setRunningMode(com.google.mediapipe.tasks.vision.core.RunningMode.IMAGE)
            .setNumFaces(1)

        faceLandmarker = FaceLandmarker.createFromOptions(context, optionsBuilder.build())
    }

    private fun resetResults() {
        val regions = listOf("forehead", "left_cheek", "right_cheek", "under_eye_left", "under_eye_right", "lips", "left_cheek_hollow", "right_cheek_hollow")
        results = mutableMapOf()
        for (region in regions) {
            results[region] = mutableMapOf("r" to mutableListOf(), "g" to mutableListOf(), "b" to mutableListOf())
        }
        geometry = mutableMapOf(
            "face_width" to mutableListOf(),
            "face_height" to mutableListOf(),
            "eye_aperture" to mutableListOf()
        )
    }

    fun startCollection() {
        resetResults()
        capturedImageBase64 = null
        frameCount = 0
        isCollecting = true
    }

    fun stopCollection(): Map<String, Any> {
        isCollecting = false
        val finalResult = mutableMapOf<String, Any>()
        finalResult.putAll(results)
        finalResult["geometry"] = geometry
        capturedImageBase64?.let { finalResult["thumbnail"] = it }
        return finalResult
    }

    fun processFrame(bitmap: Bitmap) {
        if (!isCollecting) return
        
        if (frameCount == 0) {
            captureThumbnail(bitmap)
        }
        frameCount++

        val mpImage = BitmapImageBuilder(bitmap).build()
        val result = faceLandmarker?.detect(mpImage)

        if (result != null && result.faceLandmarks().isNotEmpty()) {
            val landmarks = result.faceLandmarks()[0]
            
            // ROI抽出
            extractRGB(landmarks[10], "forehead", bitmap)
            extractRGB(landmarks[234], "left_cheek", bitmap)
            extractRGB(landmarks[454], "right_cheek", bitmap)
            extractRGB(landmarks[101], "under_eye_left", bitmap)
            extractRGB(landmarks[330], "under_eye_right", bitmap)
            extractRGB(landmarks[0], "lips", bitmap)
            extractRGB(landmarks[214], "left_cheek_hollow", bitmap)
            extractRGB(landmarks[434], "right_cheek_hollow", bitmap)

            // 幾何学データ
            val faceWidth = Math.abs(landmarks[454].x() - landmarks[234].x())
            val faceHeight = Math.abs(landmarks[152].y() - landmarks[10].y())
            val leftEyeApt = Math.abs(landmarks[159].y() - landmarks[145].y()) / Math.max(0.01f, Math.abs(landmarks[133].x() - landmarks[33].x()))
            val rightEyeApt = Math.abs(landmarks[386].y() - landmarks[374].y()) / Math.max(0.01f, Math.abs(landmarks[362].x() - landmarks[263].x()))

            geometry["face_width"]?.add(faceWidth.toDouble())
            geometry["face_height"]?.add(faceHeight.toDouble())
            geometry["eye_aperture"]?.add(((leftEyeApt + rightEyeApt) / 2.0).toDouble())

            // UI更新通知
            val update = mapOf(
                "status" to "detected",
                "landmarks" to mapOf(
                    "forehead" to mapOf("x" to landmarks[10].x(), "y" to landmarks[10].y()),
                    "left_cheek" to mapOf("x" to landmarks[234].x(), "y" to landmarks[234].y()),
                    "right_cheek" to mapOf("x" to landmarks[454].x(), "y" to landmarks[454].y()),
                    "lips" to mapOf("x" to landmarks[0].x(), "y" to landmarks[0].y())
                ),
                "geometry" to mapOf(
                    "face_width" to faceWidth,
                    "face_height" to faceHeight,
                    "eye_aperture" to (leftEyeApt + rightEyeApt) / 2.0f
                )
            )
            onFrameUpdate?.invoke(update)
        } else {
            onFrameUpdate?.invoke(mapOf("status" to "not_detected"))
        }
    }

    private fun extractRGB(landmark: com.google.mediapipe.tasks.components.containers.NormalizedLandmark, name: String, bitmap: Bitmap) {
        val x = (landmark.x() * bitmap.width).toInt()
        val y = (landmark.y() * bitmap.height).toInt()
        val roiSize = 15
        val halfSize = roiSize / 2
        
        var r = 0.0
        var g = 0.0
        var b = 0.0
        var count = 0.0

        for (dy in -halfSize until halfSize) {
            for (dx in -halfSize until halfSize) {
                val px = x + dx
                val py = y + dy
                if (px >= 0 && px < bitmap.width && py >= 0 && py < bitmap.height) {
                    val color = bitmap.getPixel(px, py)
                    r += android.graphics.Color.red(color)
                    g += android.graphics.Color.green(color)
                    b += android.graphics.Color.blue(color)
                    count += 1.0
                }
            }
        }

        if (count > 0.0) {
            results[name]?.get("r")?.add(r / count)
            results[name]?.get("g")?.add(g / count)
            results[name]?.get("b")?.add(b / count)
        }
    }

    private fun captureThumbnail(bitmap: Bitmap) {
        val targetWidth = 480
        val ratio = targetWidth.toFloat() / bitmap.width
        val targetHeight = (bitmap.height * ratio).toInt()
        val resized = Bitmap.createScaledBitmap(bitmap, targetWidth, targetHeight, true)
        val stream = ByteArrayOutputStream()
        resized.compress(Bitmap.CompressFormat.JPEG, 50, stream)
        capturedImageBase64 = Base64.getEncoder().encodeToString(stream.toByteArray())
    }
}
